src/sqlfluff/rules/L042.py

```diff
@@ -48,44 +70,281 @@ class Rule_L042(BaseRule):
+
+
+def _find_nested_subqueries(
+    select: Segments,
+) -> Generator[Tuple[str, Segments, Segments, BaseSegment], None, None]:
+    select_types = [
+        "with_compound_statement",
+        "set_expression",
+        "select_statement",
+    ]
+    from_clause = select.children().first(is_type("from_clause")).children()
+    offending_types = ["join_clause", "from_expression_element"]
+    # Match any of the types we care about
+    for this_seg in from_clause.children(is_type(*offending_types)).iterate_segments():
+        parent_type = this_seg[0].get_type()
+        # Ensure we are at the right depth (from_expression_element)
+        if not this_seg.all(is_type("from_expression_element")):
+            this_seg = this_seg.children(
+                is_type("from_expression_element"),
+            )
+
+        table_expression_el = this_seg.children(
+            is_type("table_expression"),
+        )
+
+        # Is it bracketed? If so, lint that instead.
+        bracketed_expression = table_expression_el.children(
+            is_type("bracketed"),
+        )
+        nested_select = bracketed_expression or table_expression_el
+        # If we find a child with a "problem" type, raise an issue.
+        # If not, we're fine.
+        seg = nested_select.children(is_type(*select_types))
+        if not seg:
+            # If there is no match there is no error
+            continue
+        # Type, parent_select, parent_sequence
+        yield (parent_type, select, this_seg, table_expression_el[0])
+
+
+def _get_casing_preference(root_select: Segments):
+    first_keyword = root_select.recursive_crawl("keyword", recurse_into=False).first()[
+        0
+    ]
+    if first_keyword.raw[0] == first_keyword.raw[0].lower():
+        return "LOWER"
+
+    return "UPPER"
+
+
+class _CTEChecker:
+    """Buffer CTE movements, so they can be applied in bulk without a recrawl."""
+
+    def __init__(self) -> None:
+        self.used_names: List[str] = []
+        self.ctes: List[BaseSegment] = []
+        self.name_idx = 0
+
+    def has_duplicates(self) -> bool:
+        return len(set(self.used_names)) != len(self.used_names)
+
+    def insert_cte(self, cte: CTEDefinitionSegment):
+        """Add a new CTE to the list as late as possible but before all its parents."""
+        output: List[BaseSegment] = []
+        id_seg = cte.get_identifier()
+        print(id_seg, id_seg.raw, id_seg.name)
+        cte_name = id_seg.raw
+        if id_seg.is_name("quoted_identifier"):
+            cte_name = cte_name[1:-1]
+
+        self.used_names.append(cte_name)
+        # This should still have the position markers of its true position
+        inbound_subquery = Segments(cte).children().last()
+        for el in self.ctes:
+            if cte in output:  # pragma: no cover
+                output.append(el)
+                continue
+            if _is_child(Segments(el).children().last(), inbound_subquery):
+                output.append(cte)
```

Comment:
The loop body is confusing, with the `continue` and two different `output.append()` calls. Could this be cleaned up a bit? Seems like the `continue` could be replaced using `else` or something like that. (I like `continue` better when the loop body is pretty long.)